src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -127,6 +164,55 @@ public void AddEntriesFromPreviousTable(NodeStateTable<T> previousTable, EntrySt
                 _states.Add(previousEntries);
             }
 
+            public void ModifyEntriesFromPreviousTable(NodeStateTable<T> previousTable, ImmutableArray<T> outputs)
+            {
+                // Semantics:
+                // For every slot in the previous table, we compare the new value.
+                // - Cached when the same
+                // - Modified when different
+                // - Removed when i > outputs.length
+                // - Added when i < previousTable.length
+
+                var previousEntries = previousTable._states[_states.Count];
+                var modifiedEntries = ArrayBuilder<(T item, EntryState state)>.GetInstance();
+
+                var previousEnumerator = previousEntries.GetEnumerator();
+                var outputEnumerator = outputs.GetEnumerator();
+
+                bool previousHasItems = previousEnumerator.MoveNext();
+                bool outputHasItems = outputEnumerator.MoveNext();
+
+                // cached or modified items
+                while (previousHasItems && outputHasItems)
+                {
+                    var previous = previousEnumerator.Current;
+                    var replacement = outputEnumerator.Current;
+
+                    // PROTOTYPE(source-generators): support custom IEqualityComparer
+                    var cached = EqualityComparer<T>.Default.Equals(previous.item, replacement);
+                    modifiedEntries.Add((replacement, cached ? EntryState.Cached : EntryState.Modified));
+
+                    previousHasItems = previousEnumerator.MoveNext();
+                    outputHasItems = outputEnumerator.MoveNext();
+                }
+
+                // removed
+                while (previousHasItems)
+                {
+                    modifiedEntries.Add((previousEnumerator.Current.item, EntryState.Removed));
+                    previousHasItems = previousEnumerator.MoveNext();
+                }
+
+                // added
+                while (outputHasItems)
+                {
+                    modifiedEntries.Add((outputEnumerator.Current, EntryState.Added));
+                    outputHasItems = outputEnumerator.MoveNext();
+                }
+
+                _states.Add(modifiedEntries.ToImmutableAndFree());
+            }
```

Comment:
Should `ToImmutableAndFree` below have a check for `_states.Length == 0` and if so just return `NodeStateTable<T>.Empty`